src/plugins/ucx/ucx_backend.cpp

```diff
@@ -359,27 +361,37 @@ void nixlUcxEngine::progressFunc()
     }
     pthrActiveCV.notify_one();
 
+    // Set POLLIN event on all fd so that the main loop would process them all on first iteration
+    for (auto &pfd: pollFds)
+        pfd.revents = POLLIN;
+
     while (!pthrStop) {
-        int i;
-        for(i = 0; i < noSyncIters; i++) {
-            for (auto &uw: uws)
-                uw->progress();
+        for (size_t wid = 0; wid < pollFds.size() - 1; wid++) {
+            if (!(pollFds[wid].revents & POLLIN))
+                continue;
+            pollFds[wid].revents = 0;
+
+            bool made_progress = false;
+            ucs_status_t status = UCS_INPROGRESS;
+            const auto &uw = uws[wid];
+            do {
+                while (uw->progress())
+                    made_progress = true;
+
+                status = ucp_worker_arm(uw->getWorker());
+            } while (status == UCS_ERR_BUSY);
+            NIXL_ASSERT(status == UCS_OK);
+
+            if (made_progress && !wid)
+                notifProgress();
         }
-        notifProgress();
-        // TODO: once NIXL thread infrastructure is available - move it there!!!
-
-        // {
-        //     static uint64_t cnt = 0;
-        //     if ( !(cnt % 1000000)) {
-        //         std::cout << "Progress round" << std::endl;
-        //     }
-        //     cnt++;
-        // }
-
-        /* Wait for predefined number of */
-        us_t start = getUs();
-        while( (start + pthrDelay) > getUs()) {
-            std::this_thread::yield();
+
+        while (poll(pollFds.data(), pollFds.size(), -1) <= 0) {
+            // Any other error would indicate bug in UCX
+            NIXL_ASSERT(!errno || errno == ENOMEM || errno == EINTR || errno == EAGAIN)
```

Comment:
i'd prefer NIXL_ERROR log message and retry the polling, rather than crashing the application with assert. Morevover, NIXL_ASSERT could also be disabled in release. and then we don't have err indication

Reply:
> Well I guess you can use this argument to say that any error from library calls (also, cuda) should crash the application, not sure it is the right approach. 

I already explained in my previous comment why this won't be an "any error" but an invariant violation, e.g. a bug in our design.

> also, what about the error not being checked in release mode?

Well, there is another assert version that doesn't get compiled-out in release mode which we could use.

Anyway, I'm not married to this assert so removing it now...